cluster/kind/kind.go

```diff
@@ -0,0 +1,172 @@
+)
+
+const (
+	dockerConfigEnvVar = "DOCKER_CONFIG"
+	kubeletConfigPath  = "/var/lib/kubelet/config.json"
+)
+
+var (
+	kubeletConfigPathTemplate    = "%s:" + kubeletConfigPath
+	googleFindDefaultCredentials = google.FindDefaultCredentials
+	clusterKindNameRE            = regexp.MustCompile("kind-(.*)")
+)
+
+func ClusterIsKind() (bool, error) {
+	name, err := clusterName()
+	if err != nil {
+		return false, fmt.Errorf("unable to determine cluster name: %v", err)
+	}
+	matches := clusterKindNameRE.FindStringSubmatch(name)
+	if len(matches) != 2 || matches[1] == "" {
+		return false, nil
+	}
+	return true, nil
+}
+
+func clusterName() (string, error) {
+	b, err := run.OutCommand("kubectl", "config", "current-context")
+	if err != nil {
+		return "", fmt.Errorf("unable to determine cluster information: %v", err)
+	}
+	return string(b), nil
+}
+
+func clusterKindName() (string, error) {
+	name, err := clusterName()
+	if err != nil {
+		return "", fmt.Errorf("unable to determine cluster name: %v", err)
+	}
+	matches := clusterKindNameRE.FindStringSubmatch(name)
+	if len(matches) != 2 || matches[1] == "" {
+		return "", fmt.Errorf("cluster %v is not a kind cluster", name)
+	}
+	return matches[1], nil
+}
+
+func clusterKindNodes() ([]string, error) {
+	name, err := clusterKindName()
+	if err != nil {
+		return nil, err
+	}
+	out, err := run.OutCommand("kind", "get", "nodes", "--name", name)
+	if err != nil {
+		return nil, err
+	}
+	nodes := []string{}
+	for _, n := range strings.Split(string(out), " ") {
+		nodes = append(nodes, strings.TrimSuffix(n, "\n"))
+	}
+	return nodes, nil
+}
+
+func SetupGARAccess(ctx context.Context, registries []string) error {
+	// Create a temporary dir to hold a new docker config that lacks credsStore.
+	// Then use `docker login` to store the generated credentials directly in
+	// the temporary docker config.
+	// See https://kind.sigs.k8s.io/docs/user/private-registries/#use-an-access-token
+	// for more information.
+	tempDockerDir, err := os.MkdirTemp("", "kne_kind_docker")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDockerDir)
+	originalConfig := os.Getenv(dockerConfigEnvVar)
+	defer os.Setenv(dockerConfigEnvVar, originalConfig)
+	if err := os.Setenv(dockerConfigEnvVar, tempDockerDir); err != nil {
+		return err
+	}
+	configPath := filepath.Join(tempDockerDir, "config.json")
+	if err := writeDockerConfig(configPath, registries); err != nil {
+		return err
+	}
+	creds, err := googleFindDefaultCredentials(ctx, "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		return fmt.Errorf("failed to find gcloud credentials: %v", err)
+	}
+	token, err := creds.TokenSource.Token()
+	if err != nil {
+		return fmt.Errorf("failed to get token from gcloud credentials: %v", err)
+	}
+	for _, r := range registries {
+		s := fmt.Sprintf("https://%s", r)
+		if err := run.LogCommandWithInput([]byte(token.AccessToken), "docker", "login", "-u", "oauth2accesstoken", "--password-stdin", s); err != nil {
```

Comment:
Nit: Directly call 'fmt.Sprintf("https://%s", r)' as s here?

Reply:
ack, prefer current for line length and readability